python/aibrix_kvcache/aibrix_kvcache/l2/connectors/infinistore.py

```diff
@@ -0,0 +1,258 @@
+# Copyright 2024 The Aibrix Team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# 	http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+from concurrent.futures import Executor
+from typing import List, Sequence, Tuple
+
+import infinistore
+import torch
+
+from ... import envs
+from ...common import AsyncBase
+from ...memory import MemoryRegion
+from ...status import Status, StatusCodes
+from . import Connector, ConnectorFeature, ConnectorRegisterDescriptor
+
+
+@AsyncBase.async_wrap(delete="_delete")
+class InfiniStoreConnector(Connector[bytes, torch.Tensor], AsyncBase):
+    """InfiniStore connector."""
+
+    def __init__(
+        self,
+        config: infinistore.ClientConfig,
+        key_suffix: str,
+        executor: Executor,
+    ):
+        super().__init__(executor)
+        self.config = config
+        self.key_suffix = key_suffix
+        self.conn: infinistore.InfinityConnection | None = None
+
+    @classmethod
+    def from_envs(
+        cls, conn_id: str, executor: Executor, **kwargs
+    ) -> "InfiniStoreConnector":
+        """Create a connector from environment variables."""
+        host_addr = kwargs.get(
+            "addr", envs.AIBRIX_KV_CACHE_OL_INFINISTORE_HOST_ADDR
+        )
+        service_port = kwargs.get(
+            "port", envs.AIBRIX_KV_CACHE_OL_INFINISTORE_SERVICE_PORT
+        )
+        dev_list = envs.AIBRIX_KV_CACHE_OL_INFINISTORE_VISIBLE_DEV_LIST
+        assert (
+            len(dev_list) > 0
+        ), "AIBRIX_KV_CACHE_OL_INFINISTORE_VISIBLE_DEV_LIST is empty"
+
+        num_visible_gpus = torch.cuda.device_count()
+
+        dev_list = dev_list[:num_visible_gpus]
+        assert num_visible_gpus % len(dev_list) == 0, (
+            "AIBRIX_KV_CACHE_OL_INFINISTORE_VISIBLE_DEV_LIST is not a "
+            "multiple of num. of visible GPUs"
+        )
+
+        # For InfiniStore RDMA, we need to map the GPU index to the RNIC
+        # index to support multi-GPU per RNIC. For example, if we have 8
+        # GPUs and 2 RNICs, then GPU 0 to 3 are mapped to RNIC 0, and GPU
+        # 4 to 7 are mapped to RNIC 1.
+        factor = num_visible_gpus // len(dev_list)
+        gpu_idx = torch.cuda.current_device()
+        rnic_idx = gpu_idx // factor
+        dev_name = dev_list[rnic_idx]
+
```

Comment:
TODO: add gid search logic and add `--hint-gid-index` to  `infinistore.ClientConfig` to support VNI mode, this can be done later